File: workspace/2023-05-19/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fn n -- fn )
    dup 0 = if
        drop
    else
        tuck * swap
        1 -
        _factorial
    endif
;

: factorial ( n -- fn )
    1 swap
    _factorial
;

: _show_factorials recursive ( max min -- )
    over over < if
        drop drop
    else
        dup factorial .
        1 + _show_factorials
    endif
;

: show_factorials ( n -- )
    s\" factorials:\n" type
    1 _show_factorials
    s\" \n" type
;

File: workspace/2023/2023-04-16/numerical-sequence.prolog/factorial.pl

factorial(0, 1).

factorial(N, Fn) :-
    N > 0,
    is(M, N - 1),
    factorial(M, Fm),
    is(Fn, N * Fm).

factorials(N, Xs) :-
    factorials(1, N, Xs).

factorials(Min, Max, []) :-
    Min > Max.

factorials(Min, Max, [X|Xs]) :-
    Min =< Max,
    factorial(Min, X),
    is(NextMin, Min + 1),
    factorials(NextMin, Max, Xs).

